Implement RAT Restrictions for 2G/3G and 4G

Takuto88 opened the pull request:
Uses the subscriber attributes to limit RAT types. You can restrict the RAT by using the following subscriber attributes:
    
    - rat_restriction_2g = forbidden
    - rat_restriction_3g = forbidden
    - rat_restriction_4g = forbidden
    
Useful feature to have if you use PyHSS as a HLR for 2G / 3G and HSS for 4G on the same PLMN but you want to restrict certain subscribers to a certain RAT. 

Here are both PCAPs for diameter and GSUP, that shows that it's working. The Diameter PCAP also shows a "normal" response with auth vectors. I included those to prove that I did not break the diameter part.

[pcaps.zip](https://github.com/user-attachments/files/23781571/pcaps.zip)

The AIR handling is a case of Copy-Paste-Driven-Development. For now, copying and modifying an existing code section as a response seems to be the most reasonable thing to do as it is in line with existing coding styles in that area. (That's code for I want to refactor the diameter part, but not today).

Takuto88 commented:
Still a draft PR because I am not sure if the responses in GSUP are correct and if the AIR even contains a RAT Type. That's just a guess. I looked in the TTCN3 jenkins artefacts of osmocom here:

https://jenkins.osmocom.org/jenkins/view/TTCN3/job/ttcn3-hlr-test-pyhss/

There are PCAPs in the build artifacts but those do not contain the IE. I could not find any reference in the entire osmocom project other than that the IE exists but I could never find a reference to where it is used.

I could use feedback if this is the correct approach or if in GSUP, you drop the subscriber at a later stage like UpdateLocation.

osmith42 commented on lib/gsup/controller/air.py:
```diff
@@ -56,7 +57,29 @@ async def handle_message(self, peer: IPAPeer, message: GsupMessage):
+                    .with_ie('cause', GMMCause.NO_SUIT_CELL_IN_LA)
+                    .build(),
+                )
+                return
```
How about defining an exception for this and removing the response logic below to the others?

lynxis commented:
GSUP should be similar to Reference 29.002: Table 8.1/7: MAP_UPDATE_GPRS_LOCATION.
GSUP merges Update GPRS Location and Update Location into a single call.

lynxis commented:
For 4G: Even you could already know which RAT should be used in the AIR (because it's only for 4G), it seems the spec doesn't expect this to be used for RAT checking. TS 29.272 lists "RAT not allowed" only in the ULR, but not in the AIR.

TS 29.272:
* Table 5.2.1.1.1/2: Update Location Answer
* Table 5.2.3.1.1/2: Authentication Information Answer

TS 29.272: 5.2.1.1.3 describes checking the RAT to be part of the ULR.

[TS 29.272 Rel 18.5](https://www.etsi.org/deliver/etsi_ts/129200_129299/129272/18.05.00_60/ts_129272v180500p.pdf)

Takuto88 commented:
You seem to be onto somthing @lynxis. After looking more closely, I think that the ULR might be the more appropriate place to perform this check.

Out of curiosity: The OsmoHLR must have had a similar check to reject the EUTRAN RAT I assume? Because at CCC Events, it must somehow reject 4G users if they disable 4G in the web interface.  I could not find a single reference to that in either the OsmoHLR itself as well as the Diameter to GSUP gateway. I was looking for it for reference but could not find it. Do you know where it is?

Takuto88 commented:
I've addressed all the feedback, rebased and also fixed some minor things that I've come across while building this. I'd be grateful you you guys could have another look.

osmith42 commented on lib/gsup/controller/ulr.py:
```diff
@@ -132,27 +141,40 @@ async def handle_message(self, peer: IPAPeer, message: GsupMessage):
+                subscriber_info = self._database.Get_Gsup_SubscriberInfo(imsi)
+                subscriber = self._database.Get_Subscriber(imsi=imsi, get_attributes=True)
+            except ValueError as e:
+                raise ULRError(f"Subscriber not found: {imsi}", GMMCause.IMSI_UNKNOWN) from e
```
Note: IMHO it would make sense to define a SubscriberNotFound exception and use it here instead of using the generic ValueError exception as otherwise this can mask other errors and return a misleading error message. I ran into this last week with a similar code path. This doesn't need to block the PR though since this was just moved around.

Takuto88 commented:
@osmith42: I've updated the RAT Type restriction check behavior and added the log messages as requested. Does this meet your expectations? If so, I would merge it.

CC @lynxis
